### ISSUE TITLE
Uncaught Unify exception with existentials, row variables, and an or-pattern

### ISSUE THREAD

**Issue by goldfirere** [ocaml/ocaml#11922]
The following program causes an uncaught `Unify` exception in at least 4.12.0, 4.14.0, and 5.0.0:

```ocaml
type _ t1 =
  | A : 'a -> [> `A of 'a ] t1
  | B : 'a -> [> `B of 'a ] t1

let f1 (type a) (t : [ `A of a | `B of a ] t1) = assert false

type t = T : _ t1 -> t

let f (T x) =
  match x with
  | (A _ | B _) as x -> ignore (f1 x : _)
```

A quick search didn't find this bug, but maybe I missed something. If this is new to you, too, we may be able to look into it and offer a fix.

**Comment by gasche** [comment]
It does look related to what is going on in https://github.com/ocaml/ocaml/pull/11799.

Here is a simpler repro case:

```
type _ t1 =
  | A : 'a -> [> `A of 'a ] t1
  | B : 'a -> [> `B of 'a ] t1

type t = T : _ t1 -> t

let f (T (type a) (x : a t1)) =
  match x with
  | (A _ | B _) as x -> ()
```

**Comment by goldfirere** [comment]
Yes, this does smell similar to #11799, but I think is distinct: #11799 does not describe an uncaught `Unify`, and the fix proposed in #11799 still throws the exception. But I will confer with @lpw25 (author of #11799) about this before trying to dive in myself.

**Comment by gasche** [comment]
I looked at the backtrace (`OCAMLRUNPARAM=b ocamlc -i test.ml`) and it indeed point to somewhere different, the uncaught exception happens during level-lowering of the pattern variables after type-checking the pattern. But the dynamics of or-pattern and `as` and GADTs are probably relevant.

**Comment by goldfirere** [comment]
This appears not to be affecting other people and seems low priority. Though it saddens me to say so, I'm happy to let this one expire. If someone else runs across it and/or it becomes higher priority, it can always be reopened.

**Comment by gasche** [comment]
Hm, this is a known bug with no fix, so I would be more tempted to keep it open. (I think the main reason why we haven't fixed it yet is that it looks fairly subtle, not something that can be easily fixed by looking at the backtrace and thinking locally about what is going on.)

**Comment by goldfirere** [comment]
Happy to reopen.

**Comment by smuenzel** [comment]
FWIW, it seems with 5.3.0 we get the following error:

```
Error: This pattern matches values of type $1
       but a pattern was expected which matches values of type 'a
       The type constructor $1 would escape its scope
       Type $1 is abstract because no corresponding cmi file was found in path.
```

**Comment by Octachron** [comment]
I am closing this issue as the uncaught exception is fixed by #14789, even if the underlying incompatibility between or-patterns, gadts and `as` variables is still here.
